### PR TITLE
Make Array#min/max optimization respect refined methods

### PR DESCRIPTION
**Pull request by jeremyevans**
While here, add a test for Array#min/max being redefined to
test_optimization.rb.

Fixes [Bug #18180]

### CODE REVIEW

File: vm_insnhelper.c [hunk 1]
```diff
@@ -4676,7 +4676,14 @@ vm_opt_newarray_max(rb_num_t num, const VALUE *ptr)
     }
     else {
 	VALUE ary = rb_ary_new4(num, ptr);
-	return rb_funcall(ary, idMax, 0);
+        const rb_callable_method_entry_t *me =
+            rb_callable_method_entry_with_refinements(rb_cArray, idMax, NULL);
+        if (me) {
+            return rb_vm_call0(GET_EC(), ary, idMax, 0, NULL, me, RB_NO_KEYWORDS);
```
Comment (XrXr) [inline vm_insnhelper.c:4682]: Might be better to pass in `ec` instead of using `GET_EC()` here?

File: vm_insnhelper.c [hunk 2]
```diff
@@ -4702,7 +4709,14 @@ vm_opt_newarray_min(rb_num_t num, const VALUE *ptr)
     }
     else {
 	VALUE ary = rb_ary_new4(num, ptr);
-	return rb_funcall(ary, idMin, 0);
+        const rb_callable_method_entry_t *me =
+            rb_callable_method_entry_with_refinements(rb_cArray, idMin, NULL);
+        if (me) {
+            return rb_vm_call0(GET_EC(), ary, idMin, 0, NULL, me, RB_NO_KEYWORDS);
```
Comment (XrXr) [inline vm_insnhelper.c:4715]: Ditto.

**Review by XrXr — APPROVED** [review]
Fix makes sense to me. It looks like what we do in other places in the codebase to call methods that are possibly refined.

File: vm_insnhelper.c (hunk 1 above)
Comment (jeremyevans) [reply vm_insnhelper.c]: Pushed a change for that in 4ca903fc7a.  Will commit if CI passes.

**Pull request merged**